Source/santactl/Commands/sync/SNTCommandSyncManager.m

```diff
@@ -199,48 +204,40 @@ - (void)listenForPushNotificationsWithSyncState:(SNTCommandSyncState *)syncState
 }
 
 - (void)processFCMMessage:(NSDictionary *)FCMmessage withMachineID:(NSString *)machineID {
-  NSData *messageData = [self extractMessageDataFrom:FCMmessage];
+  NSDictionary *message = [self messageFromMessageData:[self messageDataFromFCMmessage:FCMmessage]];
 
-  if (!messageData) {
+  if (!message) {
     LOGD(@"Push notification message is not in the expected format...dropping message");
     return;
   }
 
-  NSError *error;
-  NSDictionary *actionMessage = [NSJSONSerialization JSONObjectWithData:messageData
-                                                                options:0
-                                                                  error:&error];
-  if (!actionMessage) {
-    LOGD(@"Unable to parse push notification message value: %@", error);
+  NSString *action = message[kFCMActionKey];
+  if (!action) {
+    LOGD(@"Push notification message contains no action");
     return;
   }
 
   // Store the file name and hash in a cache. When the rule is actually added, use the cache
   // to build a user notification.
-  NSString *fileHash = actionMessage[@"file_hash"];
-  NSString *fileName = actionMessage[@"file_name"];
-  if (fileName.length && fileHash.length) {
-    [self.ruleSyncCache setObject:fileName forKey:fileHash];
-  }
-
-  NSString *action = actionMessage[@"action"];
-  if (action) {
-    LOGD(@"Push notification action: %@ received", action);
-  } else {
-    LOGD(@"Push notification message contains no action");
+  NSString *fileHash = message[kFCMFileHashKey];
+  NSString *fileName = message[kFCMFileNameKey];
+  if (fileName && fileHash) {
+    [self.ruleSyncCache setObject:[fileName copy] forKey:[fileHash copy]];
```

Comment:
NSCache does not copy its keys. Instead of having it add a ref count to something in this scope I just figured I would give it its own copy. I am not sure how values are handled so I copied that too!
Open to suggestions here ;)